cpp/tests/join/distinct_join_tests.cpp

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/column/column.hpp>
+#include <cudf/filling.hpp>
+#include <cudf/join.hpp>
+#include <cudf/sorting.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/column_utilities.hpp>
+#include <cudf_test/column_wrapper.hpp>
+#include <cudf_test/iterator_utilities.hpp>
+#include <cudf_test/table_utilities.hpp>
+#include <cudf_test/testing_main.hpp>
+#include <cudf_test/type_lists.hpp>
+
+#include <limits>
+#include <vector>
+
+template <typename T>
+using column_wrapper = cudf::test::fixed_width_column_wrapper<T>;
+using strcol_wrapper = cudf::test::strings_column_wrapper;
+using CVector        = std::vector<std::unique_ptr<cudf::column>>;
+using Table          = cudf::table;
+
+struct DistinctJoinTest : public cudf::test::BaseFixture {
+  void compare_to_reference(
+    cudf::table_view const& build_table,
+    cudf::table_view const& probe_table,
+    std::pair<std::unique_ptr<rmm::device_uvector<cudf::size_type>>,
+              std::unique_ptr<rmm::device_uvector<cudf::size_type>>> const& result,
+    cudf::table_view const& expected_table)
+  {
+    auto const& [build_join_indices, probe_join_indices] = result;
+
+    auto build_indices_span = cudf::device_span<cudf::size_type const>{*build_join_indices};
+    auto probe_indices_span = cudf::device_span<cudf::size_type const>{*probe_join_indices};
+
+    auto build_indices_col = cudf::column_view{build_indices_span};
+    auto probe_indices_col = cudf::column_view{probe_indices_span};
+
+    auto constexpr oob_policy = cudf::out_of_bounds_policy::DONT_CHECK;
+    auto build_result         = cudf::gather(build_table, build_indices_col, oob_policy);
+    auto probe_result         = cudf::gather(probe_table, probe_indices_col, oob_policy);
+
+    auto joined_cols = build_result->release();
+    auto right_cols  = probe_result->release();
```

Comment:
```suggestion
    auto joined_cols = cudf::gather(build_table, build_indices_col, oob_policy)->release();
    auto right_cols  = cudf::gather(probe_table, probe_indices_col, oob_policy)->release();
```
Just a suggestion. This ensures `build_result` and `probe_result` are not accidentally used after the `release()` calls.